### PR TITLE
Allow deleting node w/o docker container

### PR DESCRIPTION
**Pull request by xichen1**
Before, trying to delete a node with no active docker container

will cause ResourceNotFound error, and the node data in database

can't be removed. Now before stopping and deleting docker containers,

check if it exists or not. If not exists, delete the node from database directly.

### CODE REVIEW

File: src/api-engine/api/routes/node/views.py [hunk 1]
```diff
@@ -555,15 +555,19 @@ def destroy(self, request, pk=None):
...
                 node.status = "removing"
                 node.save()
                 if node.type == "orderer":
                     orderer_cnt = Node.objects.filter(
                         type="orderer", organization__network=node.organization.network).count()
                     if orderer_cnt == 1:
                         raise ResourceInUse
-                agent.stop()
-                res = True if agent.delete() else False
+                if agent_exist:
```
Comment (zhuyuanmao) [inline src/api-engine/api/routes/node/views.py:566]: I'd like to check the node status instead, if a node is created but does not have a docker container, the status should be `created` (need to double check this).

Comment (zhuyuanmao) [reply src/api-engine/api/routes/node/views.py:566]: Nice to have, we could sync the status in database when api-engine in the startup process.

Comment (yeasy) [reply src/api-engine/api/routes/node/views.py:566]: If  1) agent does not exist;  OR 2) agent exists, but container record does not exists in DB. 
We should return OK.
Else:
Try to issue delete operation.

Comment (yeasy) [reply src/api-engine/api/routes/node/views.py:566]: Try several times, if still with exception, then return false.
TODO: need optimization in future release.

Comment (xichen1) [reply src/api-engine/api/routes/node/views.py:566]: @yeasy @zhuyuanmao 

Could you indicate where to find the container record in DB? 

I didn't find an appropriate column in the Node table(and other tables such as Agent) to query the container record.

Comment (zhuyuanmao) [reply src/api-engine/api/routes/node/views.py:566]: @xichen1  There is a field that records the id of the container in the Node model, if one container was initialized for the node, there should be an ID that relates to the docker container. You'll be able to take this advantage to find the corresponding container record. 

https://sourcegraph.com/github.com/hyperledger/cello/-/blob/src/api-engine/api/models.py?L540-544&subtree=true

Comment (xichen1) [reply src/api-engine/api/routes/node/views.py:566]: Didn't notice this id, thank you!

File: src/api-engine/api/routes/node/views.py [hunk 2]
```diff
@@ -555,15 +555,41 @@ def destroy(self, request, pk=None):
...
+                else:
+                    # try to stop/delete container 3 times
+                    # TODO: optimize the retry logic
+                    for i in range(3):
+                        try:
+                            response = agent.stop()
+                            if response != True:
+                                res = False
+                                LOG.info("Retry to stop/delete container %d time(s).", i + 1)
+                                LOG.error("Exception when agent stops/deletes container: %s", response)
```
Comment (yeasy) [inline src/api-engine/api/routes/node/views.py:578]: Exception --> Failed?

File: src/api-engine/api/routes/node/views.py [hunk 3]
```diff
@@ -555,15 +555,41 @@ def destroy(self, request, pk=None):
...
+                    res = True
+                else:
+                    # try to stop/delete container 3 times
+                    # TODO: optimize the retry logic
+                    for i in range(3):
+                        try:
+                            response = agent.stop()
+                            if response != True:
+                                res = False
+                                LOG.info("Retry to stop/delete container %d time(s).", i + 1)
```
Comment (yeasy) [inline src/api-engine/api/routes/node/views.py:577]: Move to 573, before "try section"

File: src/api-engine/api/routes/node/views.py [hunk 4]
```diff
@@ -555,15 +555,41 @@ def destroy(self, request, pk=None):
...
+                                LOG.error("Exception when agent stops/deletes container: %s", response)
+                                continue
+                            response = agent.delete()
+                            if response != True:
+                                res = False
```
Comment (yeasy) [inline src/api-engine/api/routes/node/views.py:582]: Move this line to initialization at line 566.

**Review by yeasy — APPROVED** [review]

**Pull request merged**